Make replace sstables implementations exception safe

bhalevy opened the pull request:
This is the first phase of providing strong exception safety guarantees by the generic `compaction_backlog_tracker::replace_sstables`.

Once all compaction strategies backlog trackers' replace_sstables provide strong exception safety guarantees (i.e. they may throw an exception but must revert on error any intermediate changes they made to restore the tracker to the pre-update state).

Once this series is merged and ICS replace_sstables is also made strongly exception safe (using infrastructure from size_tiered_backlog_tracker introduced here), `compaction_backlog_tracker::replace_sstables` may allow exceptions to propagate back to the caller rather than disabling the backlog tracker on errors.

raphaelsc commented on compaction/compaction_strategy.cc:
```diff
@@ -168,26 +170,36 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
+    int64_t delta_bytes = 0;
+    auto all = _all;
+    all.reserve(_all.size() - old_ssts.size() + new_ssts.size());
+
```
I really dislike that we're managing the set of sstables in the tracker.

Backlog tracker will eventually land on compaction manager. Table doesn't have to be aware of it. 

I am planning to wire compaction strategy logic with the backlog tracker, so we can reuse the get_buckets() logic which is N log N on number of sstables. For inserting N sstables, you fall into quadratic behavior. The impact of this is reduced as N is small for STCS, and that we're doing a single contribution recalculation on every replacement.

If the tracker is no longer maintaining the sstable set, exception handling becomes simpler.

A failure to calculate backlog can be corrected in the next attempt. The problem was living with a partial sstable set, which doesn't reflect the one in replica::table. Having two sources of truth for the sstable set is asking for trouble.

I suggest we drop this series in favor of a revamp.

bhalevy replied:
Sometimes best is the enemy of good.
Improving exception safety with what we have doesn't prevent us from improving it in the future.

Also, achieving exception safety with the current design is feasible to bavkport, while I presume that the revamped code wouldn't be.

Stratigically, I'd like to proceed in smaller steps, so to establish a true, strongly exception safe framework, even if it's not optimal with respect to memory allocations and complexity, and then continue to improve it within the safe boundaries we've set.

avikivity replied:
I agree that it's not good for the backlog tracker to maintain its own sstable set. Soon we'll start to worry about whether the sets match.

Perhaps a way to enforce consistency is with an undo record. replace_sstables() returns an undo record that undoes the work, unless cancelled.

bhalevy replied:
The problem with the undo record is that if we want it to be efficient and undo the individual erases and inserts, it might throw in the undo path which should be noexcept.  For a wholesale undo that copies the input before modifying it, its efficiency would be the same as making the changes on temporary copies and making sure that applying them will not throw.

avikivity replied:
std::map has extract/merge members which can temporarily remove nodes with the ability to merge them back again. But it's a large project to apply the concept to our custom data structures, and maybe impossible as long as we use the interval map.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1598/

avikivity commented on utils/math.hh:
```diff
@@ -0,0 +1,40 @@
+constexpr double const_log() {
+    return std::log(i);
+}
+
```
I don't understand why we're adding such trivial helpers.

bhalevy replied:
dropped in v2

avikivity commented on utils/math.hh:
```diff
@@ -0,0 +1,40 @@
+
+template <unsigned base>
+inline double log_base(double x) {
+    return std::log(x) * inv_const_log<base>();
```
There's no need for such optimization. The compiler knows how to compute the logarithm of a constant.

avikivity replied:
Also, it's not helping to make the backlog tracker exception safe.

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -120,10 +120,10 @@ size_tiered_backlog_tracker::compacted_backlog(const compaction_backlog_tracker:
 }
 
-void size_tiered_backlog_tracker::refresh_sstables_backlog_contribution() {
+size_tiered_backlog_tracker::contribution size_tiered_backlog_tracker::calculate_sstables_backlog_contribution(const std::unordered_set<sstables::shared_sstable>& all, const sstables::size_tiered_compaction_strategy_options& stcs_options) {
```
Should it be const?

Perhaps even static (and receive parameters). "calculate" is an indication that it wants to be a pure function.

bhalevy replied:
dropped in v2

avikivity commented:
It still looks too complicated for what it does.

Alterantives:
 - undo records
 - copy state beforehand, restore state on exception

bhalevy commented on compaction/size_tiered_backlog_tracker.hh:
```diff
@@ -64,10 +64,14 @@
+    struct contribution {
+        double value = 0.0f;
+        std::unordered_set<sstables::shared_sstable> sstables;
+    };
```
I'll rename it to `sstables_backlog_contribution` then.
I hope it makes more sense this way.

bhalevy commented:
In v2 (https://github.com/scylladb/scylladb/commit/be279889e4025f9fb46876c111dc1da906890fe6):
* dropped "utils: add log_base helper"
* size_tiered_backlog_tracker: define struct sstables_backlog_contribution
  - rename `contrib` to `sstables_backlog_contribution`
* size_tiered_backlog_tracker: replace_sstables: provide strong exception safety guarantees
  - use deferred undo function
  - check erase and insert results for bookkeeping and for undo purposes

bhalevy commented:
v3 [e73235a](https://github.com/scylladb/scylladb/commit/e73235a598426c2aa7e566f3cebe6a03e538c1dd):
* size_tiered_backlog_tracker: update_sstables: update total_bytes only if set changed
   - refactored out of "size_tiered_backlog_tracker: replace_sstables: provide strong exception safety guarantees"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1665/

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -176,13 +176,30 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
+        clogger.warn("size_tiered_backlog_tracker: replace_sstables failed. Rolling back...");
+        for (auto& sst : undo_insert) {
+            _total_bytes -= sst->data_size();
+            _all.erase(std::move(sst));
```
I wonder if boost::icl::interval_map provides has nothrow erase. Usually erasing doesn't need to allocate, but for interval maps, it needs to invoke merging of nodes.

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -176,13 +176,30 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
+        }
+        for (auto& sst : undo_erase) {
+            _total_bytes += sst->data_size();
+            _all.insert(std::move(sst));
```
This certainly allocates.

bhalevy replied:
since this is in undo, mu assumption is that it is not likely to throw since those elements were there before they were erased, so it's similar to inserting after reserving capacity.

bhalevy replied:
Alternatively, we can copy `_all` and restore it on error.
This is simpler and more robust, yet less efficient.

avikivity replied:
I don't think we can make that assumption.

What we can do is, during the initial operation, only erase after all insertions are done. Then, if erases don't throw, then once insertions are done we won't need to roll back, and therefore won't need to re-insert erased elements.

There's still the problem that interval_map erase needs to allocate.

bhalevy replied:
> There's still the problem that interval_map erase needs to allocate.

But what prompted you about it?
I don't see it used in any of the backlog trackers.

Are you thinking about `partitioned_sstable_set`?
We're not there yet...

bhalevy replied:
> What we can do is, during the initial operation, only erase after all insertions are done. Then, if erases don't throw, then once insertions are done we won't need to roll back, and therefore won't need to re-insert erased elements.

Good idea.  I'mm make that change.

raphaelsc replied:
std::unordered_set::erase may allocate too during hashing, but I don't think it happens with lw_shared_ptr, which hashing and comparison are done on underlying ptr.

I think the simplest way is to copy _all and restore on failure. STCS doesn't have a high number of sstables anyway.

avikivity replied:
Isn't sstable_set manipulated here?

bhalevy replied:
It's a bit more complicated since `refresh_sstables_backlog_contribution` may still throw (e.g. from `get_buckets` -> `create_sstable_and_length_pairs`) and we'll need to undo the erasures in this case.
I think I'll go back to the calculate_sstables_backlog_contribution that will do the calculation without changing the tracker state, based on a provisional list which has `_all + new_ssts - old_ssts` and then applying the resulting `sstables_backlog_contribution` object will be noexcept and can be done only when all else succeeded.

raphaelsc replied:
> Isn't sstable_set manipulated here?

the tracker is essentially storing a copy of the main sstable set, so it must reflect what's stored in the table. as having two sources of truth is asking for trouble, I want the tracker to be feeded the current set, rather than having the tracker maintaining the set by itself.

raphaelsc replied:
> It's a bit more complicated since `refresh_sstables_backlog_contribution` may still throw (e.g. from `get_buckets` -> `create_sstable_and_length_pairs`) and we'll need to undo the erasures in this case. I think I'll go back to the calculate_sstables_backlog_contribution that will do the calculation without changing the tracker state, based on a provisional list which has `_all + new_ssts - old_ssts` and then applying the resulting `sstables_backlog_contribution` object will be noexcept and can be done only when all else succeeded.

Now that I put more thought on it, I think we can't really allow the set maintained by the tracker to deviate from the table's sstable set. Otherwise, future calculations will all be done on incorrect / outdated sstable set. Probably one of the reasons that Glauber just disabled the tracker on exception. 

We could store pending updates in the tracker and retry them in the future (but pushing the missing updates into a list can fail too), but that's probably too complicated for this.

raphaelsc replied:
Maybe the easiest way is to just kill the set maintained by the tracker.

bhalevy replied:
simplified in v4 by applying all changes to temporary variables instead of in place.

bhalevy replied:
> Maybe the easiest way is to just kill the set maintained by the tracker.

It's not that easy since stcs tracker is also embedded in twcs and lcs trackers, that will need to maintain that external set for stcs tracker.

I'll look into it more deeply as a followup.

bhalevy commented:
In v4 ([bffb541](https://github.com/scylladb/scylladb/commit/bffb54193320805237f84c59b9d09524edf2748a)):
* size_tiered_compaction_strategy: use std::ranges::range in get_buckets and friends
* size_tiered_backlog_tracker: refresh_sstables_backlog_contribution: pass ref to _all to get_buckets
* size_tiered_backlog_tracker: provide static calculate_sstables_backlog_contribution
* size_tiered_backlog_tracker: replace_sstables: provide strong exception safety guarantees
  - make all changes to temporary variables, passing a ref to the updated `all` set to calculate_sstables_backlog_contribution
  - move the updated temporaries to members in noexcept block

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1680/

avikivity commented on compaction/size_tiered_compaction_strategy.cc:
```diff
@@ -38,8 +38,10 @@ size_tiered_compaction_strategy_options::size_tiered_compaction_strategy_options
+requires std::same_as<std::ranges::range_value_t<Range>, sstables::shared_sstable>
 std::vector<std::pair<sstables::shared_sstable, uint64_t>>
-size_tiered_compaction_strategy::create_sstable_and_length_pairs(const std::vector<sstables::shared_sstable>& sstables) {
+size_tiered_compaction_strategy::create_sstable_and_length_pairs(const Range& sstables) {
```
What's the motivation? In general it's bad to over-template things.

bhalevy replied:
The motivation is to be able to accept a reference to either a vector or an unordered set

avikivity replied:
Well of course the motivation is to support multiple types, the question is are these necessary.

We should be careful sprinkling templates everywhere in high-level code.

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -177,25 +177,35 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
 void size_tiered_backlog_tracker::replace_sstables(const std::vector<sstables::shared_sstable>& old_ssts, const std::vector<sstables::shared_sstable>& new_ssts) {
+    auto all = _all;
+    auto total_bytes = _total_bytes;
+    all.reserve(all.size() + new_ssts.size());
```
This can cause a thousand allocations with 1TB/shard and 1GB/sstable. A thousand allocations isn't quite a stall, but it isn't great either.

Please see if we can switch to absl::flat_hash_set or node_hash_set.

avikivity replied:
(post-merge)

bhalevy commented on compaction/compaction_strategy.cc:
```diff
@@ -305,15 +303,52 @@ class time_window_backlog_tracker final : public compaction_backlog_tracker::imp
+                    if (r.old_ssts.empty()) {
+                        continue;
+                    }
+                    it = _windows.emplace(bound, size_tiered_backlog_tracker(_stcs_options)).first;
```
As the comment above says, the window is supposed to exist so there should be no reason to allocate.
We can immediately abort on internal error instead...

bhalevy replied:
changed to on_fatal_internal_error in v5, plus added an explanation why we can assume that windows will always exist at this point.

bhalevy commented on compaction/compaction_strategy.cc:
```diff
@@ -448,26 +448,38 @@ class leveled_compaction_backlog_tracker final : public compaction_backlog_track
         for (auto& sst : old_ssts) {
             auto level = sst->get_sstable_level();
-            _size_per_level[level] -= sst->data_size();
+            update_size_per_level(level, -sst->data_size());
```
Hmm, _size_per_level is sized to leveled_manifest::MAX_LEVELS in advance so the resize here is not needed.
Will drop it.

bhalevy replied:
Dropped resizing of size_per_level since `_size_per_level` is initialized to the max size.

bhalevy commented:
In v5 ([5455247](https://github.com/scylladb/scylladb/commit/5455247e67c5658a99aca36e9eb9185c262f31d4)):
- rebased
- size_tiered_compaction_strategy: use std::ranges::range in get_buckets and friends
  - documented motivation for template
- size_tiered_backlog_tracker: make log4 helper static
  - refactored out of next patch
- time_window_backlog_tracker: replace_sstables: provide strong exception safety guarantees
  - call on_fatal_internal_error instead of trying to allocate a non-existent window in undo()
  - documented why windows are assumed to always be present at undo time.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1788/

bhalevy commented:
ping review / merge

bhalevy commented:
> ping review / merge

ping^ review / merge

nyh commented:
@avikivity @raphaelsc you two seem to have already invested quite some time on reviewing this PR, please enter your verdict (approve or request changes).

avikivity commented on compaction/size_tiered_compaction_strategy.cc:
```diff
@@ -54,8 +56,10 @@ size_tiered_compaction_strategy::create_sstable_and_length_pairs(const std::vect
     return sstable_length_pairs;
```
It doesn't even solve any exception guarantees, since we can throw here.

Isn't it better to convert to a vector at the source, than to duplicate the entire function?

bhalevy replied:
But that would add an allocation.  I'm not sure what would be better, but I'll go with your more minimalistic approach and drop this patch from next version.

bhalevy replied:
`_all` copied into a vector in v6

avikivity commented on compaction/size_tiered_compaction_strategy.cc:
```diff
@@ -54,8 +56,10 @@ size_tiered_compaction_strategy::create_sstable_and_length_pairs(const std::vect
+requires std::same_as<std::ranges::range_value_t<Range>, sstables::shared_sstable>
 std::vector<std::vector<sstables::shared_sstable>>
-size_tiered_compaction_strategy::get_buckets(const std::vector<sstables::shared_sstable>& sstables, size_tiered_compaction_strategy_options options) {
+size_tiered_compaction_strategy::get_buckets(const Range& sstables, const size_tiered_compaction_strategy_options& options) {
```
Another large function duplicated.

bhalevy replied:
dropped in v6

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -136,7 +136,7 @@ void size_tiered_backlog_tracker::refresh_sstables_backlog_contribution() {
     auto threshold = newest_sst->get_schema()->min_compaction_threshold();
 
-    for (auto& bucket : size_tiered_compaction_strategy::get_buckets(boost::copy_range<std::vector<shared_sstable>>(_all), _stcs_options)) {
+    for (auto& bucket : size_tiered_compaction_strategy::get_buckets(_all, _stcs_options)) {
```
I think it's a net negative. Copying to a vector means the called function is not duplicated, and has simple iterators to work with.

avikivity replied:
Especially as _all can be a quite complicated type - compound_sstable_set of partitioned_sstable_set.

bhalevy replied:
okay, will drop this in next version

bhalevy replied:
dropped in v6

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -305,15 +304,53 @@ class time_window_backlog_tracker final : public compaction_backlog_tracker::imp
+                    on_fatal_internal_error(clogger, fmt::format("No window found when undoing replace_sstables for bound={} old_ssts={} new_ssts={}", bound, old_ssts, new_ssts));
+                }
+                auto& w = it->second;
+                w.replace_sstables(r.new_ssts, r.old_ssts);
```
But, this can throw, no? And we mustn't throw from a destructor.

avikivity replied:
Better to prepare the new state on the side and move it once all is done than attempt to undo and fall into the same trap.

bhalevy replied:
okay

bhalevy replied:
done in v6

bhalevy commented:
In v6 ([5710ec5](https://github.com/scylladb/scylladb/commit/5710ec55c2b510dfc320898967b984af2cbad16f)):
- Dropped following patches as per Avi's review comments:
  - size_tiered_compaction_strategy: use std::ranges::range in get_buckets and friends
  - size_tiered_backlog_tracker: refresh_sstables_backlog_contribution: pass ref to _all to get_buckets
- *_backlog_tracker: replace_sstables: provide strong exception safety guarantees
  - make all changes on temporary variables
  - and assign them back in noexcept block only when done

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2073/

bhalevy commented:
@scylladb/scylla-maint please review / merge

bhalevy commented:
https://github.com/orgs/scylladb/teams/scylla-maint ping, please review / merge

denesb commented:
@avikivity please check that your comment have been sufficiently addressed.

bhalevy commented:
> @avikivity please check that your comment have been sufficiently addressed.

ping @avikivity

avikivity commented:
re-reviewing

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -283,37 +284,46 @@ class time_window_backlog_tracker final : public compaction_backlog_tracker::imp
         };
         std::unordered_map<api::timestamp_type, replacement> per_window_replacement;
+        auto tmp_windows = _windows;
 
```
Could be made less stall-y by only copying affected windows, then only moving back affected windows.

But can be deferred.

bhalevy replied:
Isn't moving back a subset of the windows unsafe?
